### PR TITLE
Implement Lease Pinning in Store and Worker

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch builds on previous modifications to support the "pinning" of leadership via leases. In particular:
- The new lease _Pinner_ interface is implemented by all lease stores, but throws `NotImplemented` on all but the Raft store.
-  The lease worker handles pin and unpin operations.

## QA steps

New unit tests.

## Documentation changes

There will be an ultimate client-facing exposure of this facility, but not yet.

## Bug reference

N/A

### CODE REVIEW

File: state/leadership.go [hunk 1]
```diff
@@ -69,18 +69,18 @@ type leadershipChecker struct {
 }
 
 // LeadershipCheck is part of the leadership.Checker interface.
-func (m leadershipChecker) LeadershipCheck(applicationname, unitName string) leadership.Token {
-	token := m.checker.Token(applicationname, unitName)
+func (m leadershipChecker) LeadershipCheck(applicationName, unitName string) leadership.Token {
```
Comment (manadart) [inline state/leadership.go]: Drive-by name fixes. Apologies for the noise.

File: worker/lease/manager.go [hunk 2]
```diff
@@ -259,10 +276,12 @@ func (manager *Manager) handleClaim(claim claim) (bool, error) {
 			info, found := store.Leases()[claim.leaseKey]
 			switch {
 			case !found:
-				manager.config.Logger.Tracef("[%s] %s asked for lease %s, no lease found, claiming for %s", manager.logContext, claim.holderName, claim.leaseKey.Lease, claim.duration)
+				manager.config.Logger.Tracef("[%s] %s asked for lease %s, no lease found, claiming for %s",
```
Comment (manadart) [inline worker/lease/manager.go]: Again, just line length fixes -> noise.

File: worker/lease/manager.go [hunk 3]
```diff
@@ -424,6 +448,14 @@ func (manager *Manager) startRetry() *retry.Attempt {
 	)
 }
 
+func (manager *Manager) handlePin(p pin) {
+	p.respond(errors.Trace(manager.config.Store.PinLease(p.leaseKey)))
```
Comment (manadart) [inline worker/lease/manager.go]: These and similar calls violate the (Law of Demeter)[https://en.wikipedia.org/wiki/Law_of_Demeter], but I tried to limit how much I bit off with this patch.

**Comment by manadart** [comment]
[Implement Pinning in the Lease Store and Lease Worker](https://trello.com/c/jTvLezGZ/641-implement-pinning-in-the-lease-store-and-lease-worker)

**Comment by babbageclunk** [comment]
!!build!!

File: apiserver/params/leadership.go [hunk 4]
```diff
@@ -40,11 +40,11 @@ type ReleaseLeadershipBulkParams struct {
 // leadership claim.
 type ReleaseLeadershipParams struct {
```
Comment (babbageclunk) [inline apiserver/params/leadership.go]: I don't think this is used anywhere (and there's not a way to do it in the underlying store interface atm) so maybe just remove it?

File: worker/lease/bound.go [hunk 5]
```diff
@@ -29,11 +29,7 @@ type boundManager struct {
 
 // Claim is part of the lease.Claimer interface.
 func (b *boundManager) Claim(leaseName, holderName string, duration time.Duration) error {
-	key := lease.Key{
-		Namespace: b.namespace,
-		ModelUUID: b.modelUUID,
-		Lease:     leaseName,
-	}
+	key := b.leaseKey(leaseName)
```
Comment (babbageclunk) [inline worker/lease/bound.go]: Nice.

File: worker/lease/manager.go (hunk 3 above)
Comment (babbageclunk) [reply worker/lease/manager.go]: Maaaaaaaaybe? I don't think the config struct here should be considered to be an independent object from a Demeter point of view. The way we use them is just for structuring and brevity - they almost never have behaviour other than validation. It's not clear that the code would be better if the New func assigned those fields one-by-one and the Demeter violation was replaced by `manager.store.PinLease`.

Comment (babbageclunk) [reply worker/lease/manager.go]: I guess it depends how you want to fix it. :)

**Comment by manadart** [comment]
$$merge$$
